validate.go

```diff
@@ -16,33 +17,75 @@ func validate(payload []byte) ([]byte, error) {
 			kubewarden.Code(400))
 	}
 
-	logger.Info("validating request")
+	if len(settings.AllowedTypes) == 0 {
+		// empty AllowedType list, rejecting
+		return kubewarden.RejectRequest(
+			kubewarden.Message("No volume type is allowed"),
+			kubewarden.NoCode)
+	}
 
-	data := gjson.GetBytes(
-		payload,
-		"request.object.metadata.name")
+	if len(settings.AllowedTypes) == 1 && settings.AllowedTypes[0] == "*" {
+		// all volume types accepted
+		return kubewarden.AcceptRequest()
+	}
 
-	if !data.Exists() {
-		logger.Warn("cannot read object name from metadata: accepting request")
+	volumes := gjson.GetBytes(
+		payload,
+		"request.object.spec.volumes")
+	if !volumes.Exists() {
+		// pod defines no volumes, accepting
 		return kubewarden.AcceptRequest()
 	}
-	name := data.String()
 
-	logger.DebugWithFields("validating ingress object", func(e onelog.Entry) {
-		namespace := gjson.GetBytes(payload, "request.object.metadata.namespace").String()
+	logger.DebugWithFields("validating pod object", func(e onelog.Entry) {
+		name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+		namespace := gjson.GetBytes(payload,
+			"request.object.metadata.namespace").String()
 		e.String("name", name)
 		e.String("namespace", namespace)
 	})
 
-	if settings.DeniedNames.Contains(name) {
-		logger.InfoWithFields("rejecting ingress object", func(e onelog.Entry) {
-			e.String("name", name)
-			e.String("denied_names", settings.DeniedNames.String())
+	for _, volume := range volumes.Array() {
+		// obtain volumeName, volumeType:
+		var volumeName, volumeType string
+		volume.ForEach(func(key, value gjson.Result) bool {
+			if key.String() == "name" {
+				volumeName = value.String()
+			} else {
+				// must be the type
+				volumeType = key.String()
+			}
+			return true // keep iterating
 		})
 
+		match := false
+		for _, allowedType := range settings.AllowedTypes {
+			if allowedType == volumeType {
+				match = true
+			}
+		}
```

Comment:
This is just a matter of personal taste. The lack of Set into Go std library leads to a lot of code like that to be written.

You could bring back the set package, make `settings.AllowedTypes` a Set, and then replace this loop with a one liner (which has also better performances because how the Set is structured internally)

Reply:
Good idea! done in https://github.com/kubewarden/psp-volumes-policy/pull/1/commits/6ea8abb85e986527d5d9bdb5e6bbd4c8585d2614.